### PR TITLE
Adding JavaCC parser for COPY command.

### PR DESCRIPTION
**Pull request by Vizerai**

### CODE REVIEW

File: src/main/javacc/build.xml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  <target name="copy" description="Builds 'copy parser'">
+    <delete dir="copy"/>
+    <mkdir dir="copy"/> 
+    <copy file="copy.jjt" todir="copy"/>
+    <copy file="ASTID.java" todir="copy"/>
+    <copy file="ASTColumnElement.java" todir="copy"/>
+    <copy file="ASTCopyOptionElement.java" todir="copy"/>
+    <copy file="ASTFilename.java" todir="copy"/>
+    <copy file="CopyDumpVisitor.java" todir="copy"/>
```
Comment (snehashah16) [inline src/main/javacc/build.xml]: lets remove this if u remove the actual  file

File: src/main/javacc/CopyDumpVisitor.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
+package com.google.cloud.spanner.pgadapter.parsers;
+
+public class CopyDumpVisitor implements CopyVisitor
```
Comment (snehashah16) [inline src/main/javacc/CopyDumpVisitor.java]: discussed removing this file in 1:1

File: pom.xml [hunk 3]
```diff
@@ -89,6 +89,16 @@
       <artifactId>commons-cli</artifactId>
       <version>1.4</version>
     </dependency>
+    <dependency>
```
Comment (snehashah16) [inline pom.xml:92]: let's add unit tests when integrated with the existing code

**Review by snehashah16 — APPROVED** [review]

File: src/main/javacc/copy.jjt [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+  <O_BINARY: "BINARY">
+| <O_CSV: "CSV">
+| <O_FORMAT: "FORMAT">
+| <O_FREEZE: "FREEZE">
+| <O_DELIMITER: "DELIMITER">
+| <O_NULL: "NULL">
+| <O_HEADER: "HEADER">
+| <O_QUOTE: "QUOTE">
+| <O_ESCAPE: "ESCAPE">
+| <O_FORCE_QUOTE: "FORCE_QUOTE">
```
Comment (tinaspark) [inline src/main/javacc/copy.jjt:62]: some of these options, as well as COPY FROM <filename> won't be supported in the first version of COPY. Should we throw a parse exception here if those options are used or handle the exception later on?

Comment (Vizerai) [reply src/main/javacc/copy.jjt:62]: I added exceptions for the unused options. Later we can add support for them if necessary.

File: src/main/javacc/CopyDumpVisitor.java (hunk 2 above)
Comment (Vizerai) [reply src/main/javacc/CopyDumpVisitor.java]: Done.

**Review by voulg — APPROVED** [review]

File: src/main/javacc/build.xml (hunk 1 above)
Comment (Vizerai) [reply src/main/javacc/build.xml]: Done.

File: src/main/javacc/ASTColumnElement.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package com.google.cloud.spanner.pgadapter.parsers;
+
+public class ASTColumnElement extends SimpleNode {
```
Comment (libingye816) [inline src/main/javacc/ASTColumnElement.java:3]: Why do we manually add those AST classes? I found that there are a lot more AST classes referenced in the visitor but they are not here so why we have to treat those AST classes specially?

File: src/main/javacc/copy.jjt [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+{
+  < #DIGIT: ["0"-"9"] >
+|
+  < #LETTER: ["_","a"-"z","A"-"Z"] >
+|
+  < #LOWER_CASE_LETTER: ["_","a"-"z"] >
+|
+  < #ID: <LETTER> (<LETTER>|<DIGIT>)*>
+|
+  < IDENTIFIER: (<K_QUOTE> <ID> <K_QUOTE> | <ID>)>
```
Comment (libingye816) [inline src/main/javacc/copy.jjt:88]: Why is the identifier wrapped with a single quote? Also, why quoted identifier still has the restrictions that only digits and letters are allowed?

File: src/main/javacc/copy.jjt [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+  | {} /* EMPTY */
+}
+
+/** Column List. */
+void columnList() #ColumnList : {}
+{
+    LOOKAHEAD(2)
+    "(" columnList() ")"
+  | LOOKAHEAD(2)
+    columnElem() "," columnList()
```
Comment (libingye816) [inline src/main/javacc/copy.jjt:129]: I am trying to understand why we need this given that we already have L127. Does this imply that columns are not necessarily wrapped by parenthesis?

File: src/main/javacc/copy.jjt [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+/** Copy From. */
+void copyFrom() #CopyFrom : 
+{
+  Object o;
+}
+{
+  <K_FROM>	{o = Boolean.TRUE;}
+  | <K_TO>	{
+      o = Boolean.FALSE;
+      throw new ParseException("COPY TO is not supported.");
```
Comment (libingye816) [inline src/main/javacc/copy.jjt:154]: A little confused here. As Tina mentioned that COPY FROM is not supported, it looks like neither of them is supported.

**Review by libingye816 — COMMENTED** [review]
Not very familiar with COPY work so I have a couple of questions. Thanks for the patience.

A general question: What are the command we are going to support in this PR? I noticed that not all COPY syntax is supported. Do we plan to support them in the future?

File: src/main/javacc/copy.jjt [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+{
+  <K_WITH>
+  | {} /* EMPTY */
+}
+
+void copyOptions() #CopyOptions : {}
+{
+  LOOKAHEAD(2)
+  "(" copyOptionList() ")"
+  | copyOptionList()
```
Comment (libingye816) [inline src/main/javacc/copy.jjt:178]: I am trying to understand why we need L178 given that we already have L177. Does this imply that copyOptions are not necessarily wrapped by parenthesis?

Comment (Vizerai) [reply src/main/javacc/copy.jjt:178]: Yes, the ( ) are optional.

File: src/main/javacc/copy.jjt (hunk 8 above)
Comment (Vizerai) [reply src/main/javacc/copy.jjt:154]: The exception is only thrown on <K_TO>.

File: src/main/javacc/copy.jjt (hunk 6 above)
Comment (Vizerai) [reply src/main/javacc/copy.jjt:88]: Identifiers can be wrapped in ' ' marks.  Identifiers should also allow underscores '_'.

File: src/main/javacc/ASTColumnElement.java (hunk 5 above)
Comment (Vizerai) [reply src/main/javacc/ASTColumnElement.java:3]: I manually added ones that I wanted to give specific functionality to.

**Pull request merged**